File: 11-graph/00695-Max_Area_Of_Island/00695-max_area_of_island.java
// 695: Max Area Of Island
// https://leetcode.com/problems/max-area-of-island


class Solution {
    int result = 0;

    public int maxAreaOfIsland(int[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i == grid.length || j == grid[0].length || grid[i][j] == 0) {
            return 0;  
        }
    
        grid[i][j] = 0;
        
        return (
            1 + maxAreaOfIsland(grid, i + 1, j)
            + maxAreaOfIsland(grid, i - 1, j)
            + maxAreaOfIsland(grid, i, j + 1)
            + maxAreaOfIsland(grid, i, j - 1)
        );
    }
    
    // SOLUTION
    public int maxAreaOfIsland(int[][] grid) {
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                result = Math.max(result, maxAreaOfIsland(grid, i, j));
            }
        }
    
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[][] grid = {
            {0,0,1,0,0,0,0,1,0,0,0,0,0},
            {0,0,0,0,0,0,0,1,1,1,0,0,0},
            {0,1,1,0,1,0,0,0,0,0,0,0,0},
            {0,1,0,0,1,1,0,0,1,0,1,0,0},
            {0,1,0,0,1,1,0,0,1,1,1,0,0},
            {0,0,0,0,0,0,0,0,0,0,1,0,0},
            {0,0,0,0,0,0,0,1,1,1,0,0,0},
            {0,0,0,0,0,0,0,1,1,0,0,0,0}
        };

        // OUTPUT
        var result = o.maxAreaOfIsland(grid);
        System.out.println(result);
    }
}

File: 11-graph/00695-Max_Area_Of_Island/00695-max_area_of_island.py
# 695: Max Area Of Island
# https://leetcode.com/problems/max-area-of-island

from typing import List


class Solution:
    # SOLUTION
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        visit = set()

        def dfs(i: int, j: int):
            if (i < 0 or i == m or j < 0 or j == n or grid[i][j] == 0 or (i, j) in visit):
                return 0

            visit.add((i, j))

            return \
                1 + dfs(i + 1, j) \
                + dfs(i - 1, j) \
                + dfs(i, j + 1) \
                + dfs(i, j - 1)


        result = 0
        for r in range(m):
            for c in range(n):
                result = max(result, dfs(r, c))

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    grid = [
        [0,0,1,0,0,0,0,1,0,0,0,0,0],
        [0,0,0,0,0,0,0,1,1,1,0,0,0],
        [0,1,1,0,1,0,0,0,0,0,0,0,0],
        [0,1,0,0,1,1,0,0,1,0,1,0,0],
        [0,1,0,0,1,1,0,0,1,1,1,0,0],
        [0,0,0,0,0,0,0,0,0,0,1,0,0],
        [0,0,0,0,0,0,0,1,1,1,0,0,0],
        [0,0,0,0,0,0,0,1,1,0,0,0,0]
    ]

    # OUTPUT
    result = o.maxAreaOfIsland(grid)
    print(result)
